activation/activation_test.go

```diff
@@ -314,6 +315,46 @@ func TestBuilder_RestartSmeshing(t *testing.T) {
 	})
 }
 
+func TestBuilder_StartSmeshingAvoidsPanickingIfPrepareInitializerReturnsError(t *testing.T) {
+	// First verify that a panic occurs when PrepareInitializer does not return
+	// an error but StartSession does.
+	l := log.NewMockLogger(gomock.NewController(t))
+	l.EXPECT().Panic(gomock.Any(), gomock.Any())
+
+	tab := newTestBuilder(t)
+	tab.log = l
+
+	// Stub these methods in case they get called
+	tab.mpost.EXPECT().GenerateProof(gomock.Any(), gomock.Any()).AnyTimes()
+	tab.mclock.EXPECT().AwaitLayer(gomock.Any()).AnyTimes()
+
+	// Set expectations
+	tab.mpost.EXPECT().PrepareInitializer(gomock.Any(), gomock.Any()).Return(nil)
+	tab.mpost.EXPECT().StartSession(gomock.Any()).Return(errors.New("should panic"))
+	tab.StartSmeshing(tab.coinbase, PostSetupOpts{})
+
+	// Now verify that a panic does not occur if PrepareInitializer returns an error
+	l = log.NewMockLogger(gomock.NewController(t))
+	tab = newTestBuilder(t)
+	tab.log = l
+
+	// Stub these methods in case they get called
+	tab.mpost.EXPECT().GenerateProof(gomock.Any(), gomock.Any()).AnyTimes()
+	tab.mclock.EXPECT().AwaitLayer(gomock.Any()).AnyTimes()
+
+	// Set expectations
+	tab.mpost.EXPECT().PrepareInitializer(gomock.Any(),
+		gomock.Any()).Return(errors.New("avoid panic"))
+	// We don't expect this call to be made, but we need StartSession to return
+	// an error if it is called so that l.Panic is then called. If we didn't set this then
+	// the test could pass even if StartSession was called.
+	tab.mpost.EXPECT().StartSession(gomock.Any()).Return(errors.New("err")).MaxTimes(0)
+	tab.StartSmeshing(tab.coinbase, PostSetupOpts{})
+
+	// We wait a second since the call to l.Panic is happening in a different goroutine
+	time.Sleep(time.Second)
```

Comment:
Now it becomes philosophical 😄, either:

- we write the test knowing how the service works internally; then we don't need to sleep because we know that the panic won't be called when `PrepareInitializer` returns an error
- we write the test and tread the service as a black box; then we should make sure that no go routine is started that could call `StartSmeshing` when `PrepareInitializer` returns an error; we can do this by setting the `errGroup` to `nil` for this part of the test or mocking it and asserting that no go-routine was started

While the second is safer, I think the first is OK here - also if someone changes the method to not return when `PrepareInitializer` returns an error and a go routine is started by `StartSmeshing`, go detects this:

```
panic: Fail in goroutine after TestBuilder_StartSmeshingAvoidsPanickingIfPrepareInitializerReturnsError has completed
```

Reply:
I don't think the first approach is a valid approach, the whole point of this test is to verify that if `PrepareInitializer` returns an error then `log.Panic` is not called, if we assume that's already the case then we don't need the test.

The second approach as you describe it, is not blackbox, because you'd need to  manipulate the internal `errGroup` with an understanding that use of the `errGroup` implies that eventually `log.Panic` would be called.

So IMO the blackbox approach would be to either sleep at the end of the test or rely on the go test framework detecting a failure after the test has completed.

I prefer the sleep (_even though I really don't like having to use sleep in tests_) because it provides a much clearer error ( relying on the go test framework simply provides a stack trace which lets you know there was some problem with gomock) and 1 second is not a hugely significant amount of time to wait when our tests take more than 5 minutes to run.

Reply:
in my opinion sleeps in tests should not be used at all, relying on sleep is just bad code, regardless of reasons

Reply:
Upon further discussion @fasmat still preferred to remove the sleep and have a faster test, so that is what I have done.

Reply:
@dshulyak In this case what would you do then?